### PR TITLE
[SYCL-MLIR]: Attach attributes to function parameters and return value.

### PR DESCRIPTION
**Pull request by etiotto**
Use the new attribute infrastructure to attach attributes to function parameters and return value. 

Note: this PR builds upon #7126

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 1]
```diff
@@ -181,6 +185,56 @@ static void addNoBuiltinAttributes(mlirclang::AttrBuilder &FuncAttrs,
   llvm::for_each(NBA->builtinNames(), AddNoBuiltinAttr);
 }
 
+static bool DetermineNoUndef(QualType QT, clang::CodeGen::CodeGenTypes &Types,
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:188]: Moved here from clang-mlir.cc

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 2]
```diff
@@ -1565,6 +1619,33 @@ CodeGenTypes::arrangeGlobalDeclaration(clang::GlobalDecl GD) {
   return CGM.getTypes().arrangeGlobalDeclaration(GD);
 }
 
+bool CodeGenTypes::IsLLVMStructABI(const clang::RecordDecl *RD,
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:1622]: Moved from clang-mlir.cc

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 3]
```diff
@@ -75,142 +68,13 @@ constexpr llvm::StringLiteral MLIRASTConsumer::DeviceModuleName;
...
-mlir::IntegerAttr CodeGenUtils::wrapIntegerMemorySpace(unsigned memorySpace,
-                                                       MLIRContext *ctx) {
+static mlir::IntegerAttr wrapIntegerMemorySpace(unsigned memorySpace,
+                                                MLIRContext *ctx) {
   return memorySpace ? mlir::IntegerAttr::get(mlir::IntegerType::get(ctx, 64),
                                               memorySpace)
                      : nullptr;
 }
 
-bool CodeGenUtils::isAggregateTypeForABI(QualType qt) {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:85]: Moved to CodeGenTypes.cc

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 4]
```diff
@@ -75,142 +68,13 @@ constexpr llvm::StringLiteral MLIRASTConsumer::DeviceModuleName;
...
-
-void CodeGenUtils::TypeAndAttrs::getAttributes(
-    const SmallVectorImpl<TypeAndAttrs> &descriptors,
-    SmallVectorImpl<mlir::NamedAttrList> &attrs) {
-  assert(attrs.empty() && "Expecting 'attrs' to be empty");
-
-  for (const TypeAndAttrs &desc : descriptors)
-    attrs.push_back(desc.attrs);
-}
-
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:213]: Removed as they are no longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 5]
```diff
@@ -3086,117 +2949,6 @@ MLIRASTConsumer::getMLIRLinkage(llvm::GlobalValue::LinkageTypes LV) {
   llvm_unreachable("Unexpected linkage");
 }
 
-void MLIRASTConsumer::createMLIRParameterDescriptors(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3089]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 6]
```diff
@@ -3207,42 +2959,13 @@ MLIRASTConsumer::createMLIRFunction(const FunctionToEmit &FTE,
   const clang::CodeGen::CGFunctionInfo &FI = GetOrCreateCGFunctionInfo(&FD);
   mlir::FunctionType funcTy = getTypes().getFunctionType(FI, FD);
 
-  SmallVector<CodeGenUtils::ParmDesc, 4> parmDescriptors;
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3210]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 7]
```diff
@@ -3207,42 +2959,13 @@ MLIRASTConsumer::createMLIRFunction(const FunctionToEmit &FTE,
...
-  }
-
   mlir::FunctionOpInterface function =
       FD.hasAttr<SYCLKernelAttr>()
           ? Builder.create<gpu::GPUFuncOp>(loc, mangledName, funcTy)
           : Builder.create<func::FuncOp>(loc, mangledName, funcTy);
 
   setMLIRFunctionVisibility(function, FTE, ShouldEmit);
   setMLIRFunctionAttributes(function, FTE, ShouldEmit);
-  setMLIRFunctionParmsAttributes(function, parmDescriptors);
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3244]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 8]
```diff
@@ -3262,21 +2985,6 @@ MLIRASTConsumer::createMLIRFunction(const FunctionToEmit &FTE,
   return function;
 }
 
-void MLIRASTConsumer::createMLIRResultDescriptors(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3265]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 9]
```diff
@@ -3619,35 +3327,22 @@ void MLIRASTConsumer::setMLIRFunctionAttributes(
     PAL.addFnAttrs(attrBuilder);
   }
 
+  // Set function attributes.
   mlirclang::AttributeList FnAttrs(function->getAttrDictionary(), {}, {});
   FnAttrs.addFnAttrs(PAL.getFnAttrs(), *ctx);
   function->setAttrs(FnAttrs.getFnAttrs().getDictionary(ctx));
-}
-
-void MLIRASTConsumer::setMLIRFunctionParmsAttributes(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3627]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 10]
```diff
@@ -116,52 +116,6 @@ class FunctionToEmit {
   const FunctionContext funcContext;
 };
 
-class CodeGenUtils {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:119]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 11]
```diff
@@ -281,18 +235,6 @@ class MLIRASTConsumer : public clang::ASTConsumer {
                                                std::string mangledName,
                                                bool ShouldEmit);
 
-  /// Fill in \p parmDescriptors with the MLIR types of the \p FD function
-  /// declaration's parameters.
-  void createMLIRParameterDescriptors(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:286]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 12]
```diff
@@ -281,18 +235,6 @@ class MLIRASTConsumer : public clang::ASTConsumer {
...
-      const clang::FunctionDecl &FD,
-      llvm::SmallVectorImpl<CodeGenUtils::ParmDesc> &parmDescriptors);
-
-  /// Fill in \p resDescriptors with the MLIR types of the \p FD function
-  /// declaration's return value(s).
-  void createMLIRResultDescriptors(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:292]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 13]
```diff
@@ -301,19 +243,6 @@ class MLIRASTConsumer : public clang::ASTConsumer {
   void setMLIRFunctionAttributes(mlir::FunctionOpInterface function,
                                  const FunctionToEmit &FTE, bool ShouldEmit);
 
-  /// Set the MLIR function parameters attributes for the given \p function.
-  void setMLIRFunctionParmsAttributes(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:305]: No longer needed.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 14]
```diff
@@ -301,19 +243,6 @@ class MLIRASTConsumer : public clang::ASTConsumer {
...
-      mlir::FunctionOpInterface function,
-      const llvm::SmallVectorImpl<CodeGenUtils::ParmDesc> &parmDescriptors)
-      const;
-
-  /// Set the MLIR function result value(s) attributes for the given \p
-  /// function.
-  void setMLIRFunctionResultAttributes(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:312]: No longer needed.

**Review by whitneywhtsang — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
Do we still need https://github.com/intel/llvm/pull/7131 after this PR? @etiotto @victor-eds

**Comment by victor-eds** [comment]
> Do we still need #7131 after this PR? @etiotto @victor-eds

This should cover that. I'll just close it.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**